### ISSUE TITLE
Assign and enforce sig ownership of all e2e tests

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/kubernetes#98326]
Many e2e tests today have a `[sig-foo]` tag in their name, which lets us know which sig is responsible for fixing failing or flaky e2e tests.

However, not all tests have such a tag.  We should:
- assign `[sig-foo]` tags to all e2e tests that don't yet have them
- enforce that no new e2e tests can be added without a `[sig-foo]` tag in their name

EDIT: proposed path forward (pulled out of https://github.com/kubernetes/kubernetes/issues/98326#issuecomment-782744493 and https://github.com/kubernetes/kubernetes/issues/98326#issuecomment-783571637)
- rearrange e2e tests into sig-owned packages, 
  - `test/e2e/{sig}/...` - e2e tests that run as part of node_e2e
  - `test/e2e/common/{sig}/...` - e2e tests that run as part of e2e and node_e2e
  - add `test/e2e/common/imports.go` that imports sub-packages so e2e binaries can continue to import `test/e2e/common` to pick up test definitions ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cloud/imports.go))
  - ensure each sig-owned package has:
    - a package-local `func SIGDescribe(text string, body func()) bool` ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/framework.go))
    - an OWNERS files with sig members and `labels: sig/{sig}` ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/OWNERS)
- follow these conventions:
  - top-level `SIGDescribe` calls are required, nested `SIGDescribe` calls are not allowed
    - remove usage of `SIGFooDescribe(text)`, `SIGDescribe(sig, text)`, `KubeDescribe("[sig-foo]" + text)`, and `ginkgo.Describe("[sig-foo]")` in favor of `SIGDescribe`
  - nested `ginkgo.Describe` calls are ok, nested `SIGDescribe` calls are not
- enforce these conventions via some combination of:
  - static analysis
  - regex enforcement against names output during a `e2e.test --dry-run` run
- while we are here, `KubeDescribe` is vestigial and can be removed (ref: (https://github.com/kubernetes/kubernetes/issues/98326#issuecomment-781663980))

EDIT: Everything now conforms to policies defined above
```
$ ./hack/verify-e2e-test-ownership.sh
# ...
run at datetime: 2021-03-17T20:46:49Z
based on commit: ffb47ca3960 - 2021-03-17T16:19:25-04:00 - add verify-e2e-test-ownership.sh

[]
INFO: e2e tests in total                                : 5745
PASS: ..failing should pass all policies                : 0
PASS: ..failing must not use nested SIGDescribe         : 0
PASS: ..failing should not have multiple [sig-foo] tags : 0
PASS: ..failing must use top-level SIGDescribe          : 0
PASS: ..failing must start with [sig-foo]               : 0
PASS
```

https://github.com/kubernetes/kubernetes/pull/99698 will close this out by adding a blocking test and docs to test/e2e/README.md

/sig testing
/area test
for enforcement (FYI @BenTheElder)
/sig architecture
/area conformance
some of these are conformance tests (FYI @derekwaynecarr @dims @johnbelamaric)
/wg reliability
for helping determine appropriate sig ownership (FYI @deads2k @stevekuznetsov @wojtek-t )
/sig release
since this is relevant to figuring out who should fix release-blocking tests (FYI @kubernetes/ci-signal)
/priority important-soon
/milestone v1.21
I think it's reasonable to aim to accomplish this prior to code freeze for v1.21

**Comment by spiffxp** [comment]
/assign @spiffxp @dims @wojtek-t @justaugustus 
Assigning an arbitrary lead from each SIG/WG mentioned, please assign to someone more appropriate if you don't have the bandwidth

**Comment by aojea** [comment]
I can't see anything sig-network related :thinking: 
there are a lot of NodeConformance things though ,

**Comment by ehashman** [comment]
/assign
for visibility - lots of these are node

/sig node

**Comment by adisky** [comment]
/assign
(for sig-node)

**Comment by wojtek-t** [comment]
I've just sent out #98951 for autoscaling tests.

It seems that most of the remaining tests are node-related. We should try to fix most of these to see if there is anything that is hidden between those.

**Comment by navidshaikh** [comment]
Sent https://github.com/kubernetes/kubernetes/pull/98983 to address rest of the tests.

Echoing the review comment from #98957 

> is the [k8s.io] prefix useful once these have a sig-specific prefix?

I agree, we can remove prefix `[k8s.io]` once all the tests are tagged. @spiffxp @dims wdyt ? I'll raise a follow up to clean the prefix.

**Comment by spiffxp** [comment]
re: https://github.com/kubernetes/kubernetes/issues/98326#issuecomment-777267226

/assign @stevekuznetsov
(also maybe @ehashman knows)

at one time I thought Openshift was relying on `[k8s.io]` to filter out upstream tests for certain jobs?  I don't see that it's consistently applied some I'm assuming we could drop `[k8s.io]` from our tests? (I said as much here: https://github.com/kubernetes/kubernetes/pull/98957#discussion_r576460370)

**Comment by smarterclayton** [comment]
> at one time I thought Openshift was relying on [k8s.io] to filter out upstream tests for certain jobs?

We synthetically apply that (if the source package is in vendor/k8s.io/kubernetes) so can be dropped.

Also note in generally I am massively in favor of this change - I forced this to happen in openshift so some of the unassigned upstream tests got blanket assigned to equivalent teams (i.e. any test in k8s.io that didn't have a sig had to get one synthetically added) so this will remove the need for it.  Having owners just clarified everything, and then the ratcheting "you may not add one without" also ensured folks thought even a tiny bit about ownership.

**Comment by spiffxp** [comment]
Fleshing out what I suggested in  https://github.com/kubernetes/kubernetes/pull/98957#issuecomment-779485998 and https://github.com/kubernetes/kubernetes/pull/98957#discussion_r576460370

@wojtek-t @SergeyKanzhelev @navidshaikh @ehashman @liggitt 

Can we agree to do the following?
- rearrange e2e tests into sig-owned packages, 
  - `test/e2e/{sig}` - e2e tests that run as part of node_e2e
  - `test/e2e/common/{sig}` - e2e tests that run as part of e2e and node_e2e
- ensure each sig-owned package has:
  - a package-local `func SIGDescribe(text string, body func()) bool` ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/framework.go))
  - an OWNERS files with sig members and `labels: sig/{sig}` ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/OWNERS)
- remove usage of `SIGFooDescribe(text)`, `SIGDescribe(sig, text)`, `KubeDescribe("[sig-foo]" + text)`, and `ginkgo.Describe(...)` in favor of `SIGDescribe`
- add `test/e2e/common/imports.go` that imports sub-packages so e2e binaries can continue to import `test/e2e/common` to pick up test definitions ([example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cloud/imports.go))

Enforcing test ownership then looks like:
- an e2e test must live in a sig-owned package (or subpackage)
- an e2e test must be defined by that package's `SIGDescribe`
- PRs to that package will be assigned / auto-labeled to the appropriate SIG
- changing test ownership is a move from one package to another, requires approval from both the old SIG and new SIG

Implications I don't have the foresight to see as blockers, can you think of any?
- we disallow the concept of e2e packages containing tests owned by multiple sigs
- some moves may require imports to change or things like helper functions and fixtures to get extracted

Folks seemed to like the idea at the time but I'm seeing PRs move in different directions.  How can I help get us on the same page?

**Comment by wojtek-t** [comment]
+1 for that

I think that the PR above as an example is an intermediate step (it's not a different direction, rather just one step instead of multiple at once).

**Comment by wojtek-t** [comment]
There is one question that I have at this point - what do we do with nested Described, e.g.: this one:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/network_partition.go#L131

[there is KubeDescribe nested in SIGDescribe]

**Comment by wojtek-t** [comment]
I think that having SIGDescribe twice in those cases doesn't seem to be the best option, and those cases seem to be significant percentage of remaining cases...

**Comment by SergeyKanzhelev** [comment]
> * remove usage of `SIGFooDescribe(text)`

I don't have strong opinion here. Mostly curious why prefer the non-explicit variant? Is it to optimize for tests migrations?

**Comment by spiffxp** [comment]
>There is one question that I have at this point - what do we do with nested Described

[To ginkgo, there is no difference between `Describe`, `Context`, `When`; they're all "nested containers".](https://github.com/onsi/ginkgo/blob/dbafd0c50bd917f761b333140c58abb6d1a707f9/ginkgo_dsl.go#L322-L324)

I can't think of a strong reason to ban them, and I think it would be more trouble than it's worth to come up with alternatives for how they're used now.  I can't think of a compile-time way to enforce that "[sig-foo]" doesn't show up in their text, so I'm not sure it's worth keeping/enforcing use of  `framework` wrappers (they definitely make error reporting worse)

So something like:
- replace nested `framework.KubeDescribe` calls with `ginkgo.Describe`
- follow these conventions:
  - top-level `SIGDescribe` calls are required, nested `SIGDescribe` calls are not allowed
  - nested `ginkgo.Describe` calls are ok, nested `SIGDescribe` calls are not
- enforcement of these conventions can be done via some combination of static analysis and regex check against the full names generated by a `e2e.test --dry-run` run

**Comment by wojtek-t** [comment]
That makes sense to me.
I will start looking into migrating existing tests to satisfy these rules in free cycle, but i'm leaving the enforcement to you :)

**Comment by spiffxp** [comment]
> Mostly curious why prefer the non-explicit variant? Is it to optimize for tests migrations?

I do consider not having to modify test code to change ownership a benefit.

But it's more that we'd have to change a lot more code to conform to `SIGFooDescribe` than `SIGDescribe`.  Since we mostly already have the `SIGDescribe` pattern in place, I'd rather stick to it.
```
spiffxp@spiffxp-macbookpro:e2e (master)$ ag "SIGDescribe" -l | wc -l
     184
spiffxp@spiffxp-macbookpro:e2e (master)$ ag "SIG.+Describe" -l | wc -l
      17
```

**Comment by spiffxp** [comment]
Thinking of timeline (ref: https://github.com/kubernetes/sig-release/tree/master/releases/release-1.21#tldr)
- This work can continue past Code Freeze on Mar 9, but that's a useful milestone to evaluate our progress
- I'll update the issue description as we get tests owned
- My goal is to get enforcement landed sometime after Code Freeze
- This work can't continue past Test Freeze on Mar 24

**Comment by wojtek-t** [comment]
@spiffxp - also one more thing to do would be to ensure that no tests are tagged with more then one SIG (which I think isn't true now for a few tests)

**Comment by spiffxp** [comment]
>also one more thing to do would be to ensure that no tests are tagged with more then one SIG (which I think isn't true now for a few tests)

Good point.  There are tinier corner cases, but let's start with the big one.

/sig storage
/sig windows
/assign @msau42 @jsafrane @marosset 

There are a bunch of storage tests that look like the following:
```
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Dynamic PV (ntfs)][sig-windows] volumes should store data
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Inline-volume (ntfs)][sig-windows] volumes should allow exec of files on the volume
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Inline-volume (ntfs)][sig-windows] volumes should store data
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Pre-provisioned PV (ntfs)][sig-windows] volumes should allow exec of files on the volume
[sig-storage] In-tree Volumes [Driver: vsphere] [Testpattern: Pre-provisioned PV (ntfs)][sig-windows] volumes should store data
```
Which SIG should own these?

**Comment by marosset** [comment]
We make us of `[Feature:Windows]` for many e2e tests that only run on windows machines.
My suggestion here is probably to have sig-storage own these tests as they reside under /test/e2e/storage/* and switch to using `[Feature:Windows]` instead of `[sig-windows]` in the test names.
SIG-windows will always be happy to help debug issues here.

@jayunit100 @jingxu97 - thought?

**Comment by msau42** [comment]
I think it makes sense for storage to own the windows storage tests. Using [Feature:Windows] seems reasonable although we'll have to update a number of test jobs.

**Comment by spiffxp** [comment]
Working on in-tree enforcement before I update description.  

tl;dr 
- the majority of tests listed in the description are still not tagged (with at least one SIG), waiting on SIG Node to sign off on https://github.com/kubernetes/kubernetes/pull/99492
- usage of KubeDescribe has gone way down, but not at zero yet

**Comment by wojtek-t** [comment]
#99492 is the biggest bit now (once I get feedback from SIG node, we will be almost there with both)

**Comment by spiffxp** [comment]
refreshing description based on script in https://github.com/kubernetes/kubernetes/pull/99698

**Comment by wojtek-t** [comment]
Regarding "multi-ownership", with windows test figured (the PR didn't merge yet, but hopefully will soon), the biggest question is now about the upgrade tests.

@spiffxp @janetkuo @soltysh - there are a number of apps related upgrade tests. Even though they are currently running on GCP, I suggest that we officially make them sig-apps tests (I'm sure sig-cloud-provider-gcp will be happy to help on infrastructural problems).
Thoughts?

**Comment by soltysh** [comment]
@wojtek-t can you be more specific about which particular tests? I quickly checked https://github.com/kubernetes/kubernetes/tree/master/test/e2e/upgrades/apps/ and those are already marked as sig-apps. I don't have any objections as long as they are part of sig-apps territory but I'd like to know a concrete list before saying so :wink:

**Comment by wojtek-t** [comment]
@soltysh - sorry, I should have been more specific.
Many of those tests are actually being run from here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cloud/gcp/cluster_upgrade.go
and the tests there are marked as [sig-cloud-provider-gcp]

As part of ensuring that each test has a one single sig ownership, I suggest making it explicit that all of those are sig-apps.

Does that make more sense now?

**Comment by soltysh** [comment]
@wojtek-t most - yes, but I see a few exceptions: 

upgradeTest:
- ServiceUpgradeTest - seems like sig-network
- SecretUpgradeTest - looks like assigned to sig-storage/sig-apimachinery
- ConfigMapUpgradeTest - looks like assigned to sig-storage/sig-apimachinery
- HPAUpgradeTest - sig-autoscaling
- PersistentVolumeUpgradeTest - sig-storage
- AppArmorUpgradeTest - sig-auth (?)
- VolumeModeDowngradeTest - sig-storage

gpuUpgradeTests - looks like sig-node/sig-scheduling

kubeProxyUpgradeTests - both sig-network

kubeProxyDowngradeTests - both sig-network

serviceaccountAdmissionControllerMigrationTests - looks like sig-apimachinery

the remaining less than half :+1: for sig-apps.

**Comment by wojtek-t** [comment]
Sure - I didn't mean to put all of them to sig-apps.
There is just the biggest group of apps related one so wanted to start with this group.

**Comment by soltysh** [comment]
I'm just making sure :wink: Go for it and thx :100:

**Comment by desponda** [comment]
👋 Hello!  The code freeze is starting March 9th, 2021 (about 1 week from now). As the Issue is tagged for 1.21, is it still planned for this release?

**Comment by spiffxp** [comment]
Yes this is planned for this release, and PRs related to this are not subject to code freeze.  This will be completed by test freeze.

**Comment by wojtek-t** [comment]
@spiffxp - calling your script from head, I have:
```
INFO: e2e tests in total                                : 5732
WARN: e2e tests passing all policies                    : 5722
PASS: e2e tests should start with [sig-foo]             : 0   
WARN: e2e tests should use top-level SIGDescribe        : 1   
WARN: e2e tests should not have multiple [sig-foo] tags : 9   
PASS: e2e tests should not use nested SIGDescribe       : 0   
PASS: e2e tests should not have [k8s.io] in test names  : 0   
PASS: e2e tests should not have any KubeDescribe calls  : 0
```

So there are just couple exceptions now (and a mess with upgrade tests :) )

**Comment by wojtek-t** [comment]
@spiffxp - when you have time, there are couple PRs waiting for you now

**Comment by spiffxp** [comment]
Not reviewing on the weekend, but will check them out first thing next week. Also ok if someone else gets to them before me

**Comment by pacoxu** [comment]
```
INFO: e2e tests in total                                : 5744
WARN: ..failing should pass all policies                : 13
FAIL: ..failing must use top-level SIGDescribe          : 11
FAIL: ..failing must start with [sig-foo]               : 9
WARN: ..failing should not have multiple [sig-foo] tags : 2 
PASS: ..failing must not use nested SIGDescribe         : 0
FAIL
```
Run in master.

The `WARN: ..failing should not have multiple [sig-foo] tags` is due to https://github.com/kubernetes/kubernetes/issues/98326#issuecomment-790568639

The `FAIL: ..failing must use top-level SIGDescribe` are all from https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cloud/gcp/cluster_upgrade.go#L210

It seems it is done.

**Comment by wojtek-t** [comment]
The two opened PRs + the one that I have internally (close to ready, though dependent on the firs from the above) should actually close this.

**Comment by spiffxp** [comment]
https://github.com/kubernetes/kubernetes/pull/99698 will close this out by adding a blocking test and docs to test/e2e/README.md

**Comment by dims** [comment]
Nice work @spiffxp ... that jq stuff is something :)
